### PR TITLE
Support multiple profiles in windows_firewall_rule

### PR DESCRIPTION
**Pull request by Happycoil**
## Description
Adds Array support to the `profile` property

## Related Issue
https://github.com/chef/chef/issues/8808

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

**Comment by Happycoil** [comment]
There's a potential issue with sorting the profile names here, that means the resource will always converge unless the user also sorts the property correctly, right?

File: lib/chef/resource/windows_firewall_rule.rb [hunk 1]
```diff
@@ -158,7 +165,7 @@ def firewall_command(cmdlet_type)
           cmd << " -Direction '#{new_resource.direction}'" if new_resource.direction
           cmd << " -Protocol '#{new_resource.protocol}'" if new_resource.protocol
           cmd << " -Action '#{new_resource.firewall_action}'" if new_resource.firewall_action
-          cmd << " -Profile '#{new_resource.profile}'" if new_resource.profile
+          cmd << " -Profile '#{new_resource.profile.join("', '")}'" if new_resource.profile
```
Comment (robbkidd) [inline lib/chef/resource/windows_firewall_rule.rb]: I have not used this command in anger, so I looked up [the documentation for New-NetFirewallRule](https://docs.microsoft.com/en-us/powershell/module/netsecurity/new-netfirewallrule?view=win10-ps#parameters). It states for the `-Profile` parameter:

> Separate multiple entries with a comma and do not include any spaces.

Have you seen this change work on a Windows host? I see that the unit test was updated (and so passes) to check that the generated command includes `-Profile 'private', 'public'`, but will that space cause an error at run time?

File: lib/chef/resource/windows_firewall_rule.rb [hunk 2]
```diff
@@ -67,11 +67,18 @@ class WindowsFirewallRule < Chef::Resource
...
-        coerce: proc { |p| p.is_a?(String) ? p.downcase.to_sym : p }
-
+        coerce: proc { |p|
+          if p.is_a?(Array)
+            p.map(&:downcase).map(&:to_sym).sort
+          elsif p.is_a?(String)
+            Array(p.downcase.to_sym).sort
+          else
+            Array(p).sort
+          end
```
Comment (lamont-granquist) [inline lib/chef/resource/windows_firewall_rule.rb]: `coerce: proc { |p| Array(p).map(&:downcase).map(&:to_sym).sort }` will probably do this without the conditional.

File: lib/chef/resource/windows_firewall_rule.rb [hunk 3]
```diff
@@ -67,11 +67,18 @@ class WindowsFirewallRule < Chef::Resource
         description: "The action of the firewall rule.",
         coerce: proc { |f| f.is_a?(String) ? f.downcase.to_sym : f }
 
-      property :profile, [Symbol, String],
-        default: :any, equal_to: %i{public private domain any notapplicable},
```
Comment (lamont-granquist) [inline lib/chef/resource/windows_firewall_rule.rb]: you should still validate that the array does not contain anything that isn't one of these symbols?

File: lib/chef/resource/windows_firewall_rule.rb (hunk 1 above)
Comment (Happycoil) [reply lib/chef/resource/windows_firewall_rule.rb]: Didn't get that far yet, but it appears you're right. If they've written the cmdlet like that it's definitely not idiomatic PowerShell... I was about to say that this explains why LocalPort and RemotePort are written that way in the resource, but those parameters are actually documented as taking `String[]`

File: lib/chef/resource/windows_firewall_rule.rb (hunk 2 above)
Comment (Happycoil) [reply lib/chef/resource/windows_firewall_rule.rb]: Yeah, that's much better

File: lib/chef/resource/windows_firewall_rule.rb (hunk 3 above)
Comment (Happycoil) [reply lib/chef/resource/windows_firewall_rule.rb]: How though? I made it work with the below snippet, but is there a simpler approach?
```ruby
        callbacks: {
                      'contains values not in :public, :private :domain, :any or :notapplicable' => lambda {
                        |p| %i{public private domain any notapplicable} & p == p
                      }
                   }
```

File: lib/chef/resource/windows_firewall_rule.rb (hunk 1 above)
Comment (Happycoil) [reply lib/chef/resource/windows_firewall_rule.rb]: So turns out both approaches work for `Profile`, whereas the comma delimited string the resource uses for `LocalPort` and `RemotePort` actually doesn't work.

**Comment by Happycoil** [comment]
Not all review comments are addressed, but I'm happy with the PR when they're resolved @robbkidd @lamont-granquist

File: lib/chef/resource/windows_firewall_rule.rb (hunk 3 above)
Comment (lamont-granquist) [reply lib/chef/resource/windows_firewall_rule.rb]: that may be overly terse.  `lambda { |p| p.all? { |e| %i{public private domain any notapplicable}.include?(e) } }` or something like that which doesn't require realizing those are set operators.

File: spec/unit/resource/windows_firewall_rule_spec.rb [hunk 4]
```diff
@@ -270,6 +302,11 @@
         expect(provider.firewall_command("New")).to eql("New-NetFirewallRule -Name 'test_rule' -DisplayName 'test_rule' -Description 'Firewall rule' -Direction 'inbound' -Protocol 'TCP' -Action 'allow' -Profile 'private' -InterfaceType 'any' -Enabled 'true'")
       end
 
+      it "sets multiple Profiles" do
```
Comment (robbkidd) [inline spec/unit/resource/windows_firewall_rule_spec.rb]:
```suggestion
      it "sets multiple Profiles (must be comma-plus-space delimited for PowerShell to treat as an array)" do
```

File: spec/unit/resource/windows_firewall_rule_spec.rb [hunk 5]
```diff
@@ -326,7 +363,17 @@
...
-        expect(provider.firewall_command("Set")).to eql("Set-NetFirewallRule -Name 'test_rule' -Description 'Firewall rule' -LocalPort 80 -Direction 'inbound' -Protocol 'TCP' -Action 'allow' -Profile 'any' -InterfaceType 'any' -Enabled 'true'")
+        expect(provider.firewall_command("Set")).to eql("Set-NetFirewallRule -Name 'test_rule' -Description 'Firewall rule' -LocalPort '80' -Direction 'inbound' -Protocol 'TCP' -Action 'allow' -Profile 'any' -InterfaceType 'any' -Enabled 'true'")
+      end
+
+      it "sets LocalPort with int" do
+        resource.local_port(80)
+        expect(provider.firewall_command("Set")).to eql("Set-NetFirewallRule -Name 'test_rule' -Description 'Firewall rule' -LocalPort '80' -Direction 'inbound' -Protocol 'TCP' -Action 'allow' -Profile 'any' -InterfaceType 'any' -Enabled 'true'")
+      end
+
+      it "sets multiple LocalPorts" do
```
Comment (robbkidd) [inline spec/unit/resource/windows_firewall_rule_spec.rb]:
```suggestion
      it "sets multiple LocalPorts (must be comma-plus-space delimited for PowerShell to treat as an array)" do
```

File: spec/unit/resource/windows_firewall_rule_spec.rb [hunk 6]
```diff
@@ -336,7 +383,17 @@
...
-        expect(provider.firewall_command("Set")).to eql("Set-NetFirewallRule -Name 'test_rule' -Description 'Firewall rule' -RemotePort 443 -Direction 'inbound' -Protocol 'TCP' -Action 'allow' -Profile 'any' -InterfaceType 'any' -Enabled 'true'")
+        expect(provider.firewall_command("Set")).to eql("Set-NetFirewallRule -Name 'test_rule' -Description 'Firewall rule' -RemotePort '443' -Direction 'inbound' -Protocol 'TCP' -Action 'allow' -Profile 'any' -InterfaceType 'any' -Enabled 'true'")
+      end
+
+      it "sets RemotePort with int" do
+        resource.remote_port(443)
+        expect(provider.firewall_command("Set")).to eql("Set-NetFirewallRule -Name 'test_rule' -Description 'Firewall rule' -RemotePort '443' -Direction 'inbound' -Protocol 'TCP' -Action 'allow' -Profile 'any' -InterfaceType 'any' -Enabled 'true'")
+      end
+
+      it "sets multiple RemotePorts" do
```
Comment (robbkidd) [inline spec/unit/resource/windows_firewall_rule_spec.rb]:
```suggestion
      it "sets multiple RemotePorts (must be comma-plus-space delimited for PowerShell to treat as an array)" do
```

File: lib/chef/resource/windows_firewall_rule.rb (hunk 1 above)
Comment (robbkidd) [reply lib/chef/resource/windows_firewall_rule.rb]: Well, thanks for finding that out and fixing them!

**Pull request merged**